src/marks/auto.js

```diff
@@ -5,16 +5,31 @@ import {dot} from "./dot.js";
 import {line, lineX, lineY} from "./line.js";
 import {ruleX, ruleY} from "./rule.js";
 import {barX, barY} from "./bar.js";
-import {rect, rectX, rectY} from "./rect.js";
+import {rectX, rectY} from "./rect.js";
 import {cell} from "./cell.js";
 import {frame} from "./frame.js";
 import {bin, binX, binY} from "../transforms/bin.js";
 import {group, groupX, groupY} from "../transforms/group.js";
 import {marks} from "../mark.js";
 
 export function autoSpec(data, options) {
+  const {x, y, fx, fy, color, size, mark} = autoImpl(data, options);
+  return {x, y, fx, fy, color, size, mark};
+}
+
+function autoImpl(data, options) {
   options = normalizeOptions(options);
 
+  // Greedily materialize columns for type inference; we’ll need them anyway to
+  // plot! Note that we don’t apply any type inference to the fx and fy
+  // channels, if present; these are always ordinal (at least for now).
+  const {x, y, color, size} = options;
+  const X = materializeValue(data, x);
+  const Y = materializeValue(data, y);
+  const C = materializeValue(data, color);
+  const S = materializeValue(data, size);
```

Comment:
moved here from top of auto, which it was passing into autoSpec

Reply:
I think we still want to do this materialization (also) inside of auto, so that it doesn’t need to be done twice (the second time being when we call auto.plot).

Reply:
Ooh right, we're not passing these materialized values back out of autoImpl… and we don't want to, because autoSpec shouldn't return the materialized values. So I guess auto should greedily materialize and autoImpl shouldn't bc autoSpec should return something cleaner if possible?

Reply:
I was wrong; it doesn’t materialize twice because autoImpl returns the full markOptions with the already-materialized values. I think we’re good as-is. Nice!